### ISSUE TITLE
Allow taking (element) screenshots in testharness tests

### ISSUE THREAD

**Issue by jgraham** [web-platform-tests/wpt#14360]
Apparently mochitest has some facility for taking screenshots of elements (including iframes) which is used for some tests that want to implment something like a reftest but with more complex tests on the rendered pixel values (this was stated as a reason for not porting some tests to wpt). It seems like this should be fairly easy to implement in testdriver by exposing the screenshot primitives in WebDriver.

**Comment by jgraham** [comment]
The existing API is https://searchfox.org/mozilla-central/source/testing/mochitest/tests/SimpleTest/WindowSnapshot.js

**Comment by dholbert** [comment]
> Looking through the current uses in mozilla-central, a common pattern seems to be to run some JS, compare screenshots, run more JS, compare screenshots, etc.

Right. Some other patterns that I recall for our dependency on WindowSnapshot.js:
 - Tests for features that have a "short but unobservable" delay before the final rendering (e.g. waiting for an animated GIF to reach its final frame, to test that we do in fact play the animation).  We can't use a reftest because we don't know how long to wait, since there's no DOM event that fires when the test reaches its ready-to-snaphsot rendering.  (We could e.g. wait 5 seconds, but that's wasteful of compute time) -- so instead we "poll" the image until it matches (or fail if it never matches). Example: https://searchfox.org/mozilla-central/source/image/test/mochitest/test_animSVGImage.html, https://searchfox.org/mozilla-central/source/image/test/mochitest/test_svg_filter_animation.html
 - Tests for "visited styling", which require us to dynamically "visit" links to change the coloring, and which also may require "polling" (as above) due to having a slight delay before the visitedness is recognized via the history service. E.g. https://searchfox.org/mozilla-central/source/layout/style/test/test_visited_reftests.html#9
 - Cases where we want to test the same DOM (or nearly the same DOM) with many different tweaks. If we did this with reftests, we'd need to copypaste the same HTML into many many files (or repeat it many times in the same file); but with WindowSnapshot.js, we can have one copy and a nice compact list of our tweaks encoded in JS. Example: https://searchfox.org/mozilla-central/source/layout/style/test/test_flexbox_order.html (which tests 13 different combinations of CSS `order` values across some flex items... there may be better examples but this is the one I've got handy).

> I wonder if we could implement reftest variants

I don't know what reftest variants would look like exactly, but it'd probably be worth considering these ^^ use-cases when considering/designing them.

**Comment by jgraham** [comment]
Reftest variants would look like

```
<meta name=variant content="?option=1">
<meta name=variant content="?option=2">
```

Which would end up in the URL so you would get `/path/to/test?option=1` and `path/to/test?option=2` as separate tests. So this would possibly help with the third case above since you could write one base test and use the variant to switch the DOM however you want, but not with the other use cases. Also supporting different refs in variants would be non-trivial which could be a fairly substantial constraint.

**Comment by jugglinmike** [comment]
Could we support the first and second use case with something like `<html class="reftest-poll">`? @dholbert only mentioned the need to wait for a match, but this could be interpreted according to whatever passing conditions the test defines.

**Comment by dholbert** [comment]
> Could we support the first and second use case with something like <html class="reftest-poll">? 

That sounds good to me.
